Apps/TimelineApp/core/sensor/src/main/java/uk/ac/cam/cares/jps/sensor/source/database/SensorLocalSource.java

```diff
@@ -338,7 +374,55 @@ private List<Long> extractTimes(List<? extends SensorData> sensorDataList) {
         return times;
     }
 
+    /**
+     * Computes an MD5 hash of the given string.
+     *
+     * @param data The input string to hash.
+     * @return The MD5 hash as a hexadecimal string.
+     */
+    public String computeHash(String data) {
+        try {
+            MessageDigest digest = MessageDigest.getInstance("MD5");
+            byte[] hash = digest.digest(data.getBytes(StandardCharsets.UTF_8));
+            // Convert the byte array into a hexadecimal string
+            StringBuilder hexString = new StringBuilder();
+            for (byte b : hash) {
+                hexString.append(String.format("%02x", b));
+            }
+            return hexString.toString();
+        } catch (NoSuchAlgorithmException e) {
+            LOGGER.error("MD5 algorithm not found", e);
+            return null;
+        }
+    }
+
+    /**
+     * Checks if the given data already exists in the UnsentData table.
+     *
+     * @param sensorData The sensor data as a JSONArray.
+     * @return True if data exists, false otherwise.
+     */
+    public boolean isDataInUnsentData(JSONArray sensorData) {
```

Comment:
can modify the input to take the hash string, because it is hashed in NetworkSource anyway